test/file.jl

```diff
@@ -1156,3 +1156,70 @@ if !Sys.iswindows()
         test_22566()
     end
 end  # !Sys.iswindows
+
+function file_prefix_test(path, prefix, pfxlen=length(prefix), fails=false)
+    if Sys.iswindows() && (pfxlen > 3)
+        pfxlen = 3
+    end
+
+    filename = basename(path)
+    filepart = filename[1:pfxlen]
+    prefixpart = prefix[1:pfxlen]
+
+    if !fails
+        @test filepart == prefixpart
+    else
+        @test filepart != prefixpart
+    end
+end
+
+function test_22922()
+    def_prefix="jl_"
+    tst_prefix="ABCDEF"
+    mktempdir() do tmpdir
+        file_prefix_test(tmpdir, def_prefix)
+    end
+    mktempdir(; prefix=tst_prefix) do tmpdir
+        file_prefix_test(tmpdir, tst_prefix)
+    end
+    mktemp() do tmp,io
+        file_prefix_test(tmp, def_prefix)
+    end
+    mktemp(; prefix=tst_prefix) do tmp,io
+        file_prefix_test(tmp, tst_prefix)
+    end
+    file_prefix_test(tempname(), def_prefix)
+    #Unix like OS have prefix size limit of 5 bytes
+    file_prefix_test(tempname(; prefix=tst_prefix), tst_prefix, 5)
+
+    #Special character prefix tests
+    tst_prefix="#!@%^&*()"
+    mktempdir(; prefix=tst_prefix) do tmpdir
+        file_prefix_test(tmpdir, tst_prefix)
+    end
+    mktemp(; prefix=tst_prefix) do tmp,io
+        file_prefix_test(tmp, tst_prefix)
+    end
+
+    #Unicode test
+    tst_prefix="\u2200x\u2203y"
+    mktempdir(; prefix=tst_prefix) do tmpdir
+        file_prefix_test(tmpdir, tst_prefix)
+    end
+    mktemp(; prefix=tst_prefix) do tmp,io
+        file_prefix_test(tmp, tst_prefix)
+    end
+    #In unix the prefix is computed upto 5 bytes and not chars so will
+    #fail on unicode test. But extracting first 5 bytes and creating an
+    #invalid 5-byte character array and converting to a string will pass
+    if !Sys.iswindows()
+        bytepfx=transcode(UInt8, tst_prefix)
```

Comment:
I beg to defer here. What you claim here is accurate if the functions under discussion would have been system calls. Because you can do no better than a system offers within the purview of the system. I do not think the functions in discussion here are system calls. You can refer to system calls for Linux at:

http://syscalls.kernelgrok.com/

original Unix system calls are:

http://www.di.uevora.pt/~lmr/syscalls.html 

mkdir is a system call not mktemp to best of my knowledge. So emulating behavior of mkdir is not an accurate expectation. For performance reasons syscalls do not do input sanity checks, does not mean a user realm function should not.  

Hence, as a library you can always decide what your library should allow or not allow. Of course, your library cannot do better than what mkdir() offers. Now the question is do you consider Julia as a platform or underlying OS. In terms of design philosophy you can choose to offer whether consistency across OS is relevant or exposing the platforms internal API to the user by stating read a specific API documentation for details.

Coming to the specific use case here:

One can always argue why restrict mktemp() to behave completely like mkstemp() while mkstemp() will support suffixes as well. 

Hence, my take will be to decide on a platform behavior which will be consistent to best of Julia's currently running platforms and implement that till you reach a platform which has hypothetical prefix restrictions. The restrictions are only hypothetical, unless those limitations are implemented in the kernel with system calls for a platform. Glibc or libc behaviors can always be ignored or overwritten with your own methods.